components/director/internal/formationmapping/auth_middleware.go

```diff
@@ -0,0 +1,317 @@
+	if err != nil {
+		return false, http.StatusInternalServerError, errors.Wrap(err, "Unable to establish connection with database")
+	}
+	defer a.transact.RollbackUnlessCommitted(ctx, tx)
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	fa, err := a.faService.GetGlobalByID(ctx, formationAssignmentID)
+	if err != nil {
+		return false, http.StatusInternalServerError, errors.Wrapf(err, "while getting formation assignment with ID: %q globally", formationAssignmentID)
+	}
+
+	consumerTenantPair, err := tenant.LoadTenantPairFromContext(ctx)
+	if err != nil {
+		return false, http.StatusInternalServerError, errors.Wrap(err, "while loading tenant pair from context")
+	}
+	consumerInternalTenantID := consumerTenantPair.InternalID
+	consumerExternalTenantID := consumerTenantPair.ExternalID
+
+	log.C(ctx).Infof("Consumer with internal ID: %q and external ID: %q with type: %q is trying to update formation assignment with ID: %q for formation with ID: %q about source: %q and source type: %q, and target: %q and target type: %q", consumerInternalTenantID, consumerExternalTenantID, consumerType, fa.ID, fa.FormationID, fa.Source, fa.SourceType, fa.Target, fa.TargetType)
+	if fa.TargetType == model.FormationAssignmentTypeApplication {
+		log.C(ctx).Infof("The formation assignment that is being update has type: %s and ID: %q", model.FormationAssignmentTypeApplication, fa.Target)
+
+		app, err := a.appRepo.GetGlobalByID(ctx, fa.Target)
+		if err != nil {
+			return false, http.StatusInternalServerError, errors.Wrapf(err, "while getting application with ID: %q globally", fa.Target)
+		}
+		log.C(ctx).Infof("Successfully got application with ID: %q globally", fa.Target)
+
+		// If the consumer is integration system validate the formation assignment type is application that can be managed by the integration system caller
+		if consumerType == consumer.IntegrationSystem && app.IntegrationSystemID != nil && *app.IntegrationSystemID == consumerID {
+			log.C(ctx).Infof("The caller with ID: %q and type: %q has owner access to the formation assignment with target: %q and target type: %q that is being updated", consumerID, consumerType, fa.Target, fa.TargetType)
+			return true, http.StatusOK, nil
+		}
+
+		// Verify if the caller has owner access to the formation assignment with type application that is being updated
+		log.C(ctx).Infof("Getting parent tenant of the caller with ID: %q and type: %q", consumerID, consumerType)
+		tnt, err := a.tenantRepo.Get(ctx, consumerInternalTenantID)
+		if err != nil {
+			log.C(ctx).Warningf("an error occurred while getting tenant from ID: %q", consumerInternalTenantID)
+		}
+
+		if tnt != nil && tnt.Parent != "" {
+			exists, err := a.appRepo.OwnerExists(ctx, tnt.Parent, fa.Target)
+			if err != nil {
+				log.C(ctx).Warningf("an error occurred while verifying caller with ID: %q and type: %q has owner access to formation assignment with type: %q and target ID: %q", consumerInternalTenantID, consumerType, fa.TargetType, fa.Target)
```

Comment:
missing return in case of db error

Reply:
This is also intentional. The whole idea here is to try to validate the consumer among different scenarios and if one of these scenarios fail we don't want to interrupt the validation and return but rather abstain and continue to the other checks.

Reply:
I understand, but in the case of an error with the connection to database, isn't it correct to return the error? Logging in the case of non found is ok.

Reply:
see the comment above